source/MRMesh/MRVoxelsVolumeAccess.h

```diff
@@ -216,20 +164,23 @@ class VoxelsVolumeCachingAccessor
     void preloadNextLayer()
     {
         z_ += 1;
-        for ( auto i = 0, j = 1; j < layers_.size(); ++i, ++j )
-            std::swap( layers_[i], layers_[j] );
+        for ( auto i = 0; i + 1 < layers_.size(); ++i )
+        {
+            std::swap( layers_[i], layers_[i + 1] );
+            firstLayerVoxelId_[i] = firstLayerVoxelId_[i + 1];
+        }
         if ( z_ + params_.preloadedLayerCount - 1 < indexer_.dims().z )
             preloadLayer_( params_.preloadedLayerCount - 1 );
     }
 
     /// get voxel volume data
-    ValueType get( const Vector3i& pos ) const
+    ValueType get( const VoxelLocation & loc ) const
     {
-        const auto layerIndex = pos.z - z_;
-        if ( 0 <= layerIndex && layerIndex < layers_.size() )
-            return layers_[layerIndex][toLayerIndex( pos )];
-
-        return accessor_.get( pos );
+        const auto layerIndex = loc.pos.z - z_;
+        assert( 0 <= layerIndex && layerIndex < layers_.size() );
```

Comment:
Update the documentation to clarify the new restrictions.

Reply:
ok